File: apps/browser-demo/src/deviceAgent/__tests__/citationFilter.test.ts
// Pins the citation filter that drops blog/news subdomain citations on pricing
// questions. Without this filter, OpenAI's web_search_preview consistently surfaces
// blog.heliummobile.com posts and the model cites stale ($0 discontinued plan) prices.

import { describe, expect, it } from 'vitest';

import { filterLowAuthorityCitations, isPricingInstruction } from '../provider/citationFilter.js';

describe('isPricingInstruction', () => {
  it('matches price/cost/fee/rate/plan/subscription/monthly keywords', () => {
    expect(isPricingInstruction('what is the price of helium mobile')).toBe(true);
    expect(isPricingInstruction('check the monthly cost')).toBe(true);
    expect(isPricingInstruction('lowest fee')).toBe(true);
    expect(isPricingInstruction('current rate')).toBe(true);
    expect(isPricingInstruction('cheapest plan')).toBe(true);
    expect(isPricingInstruction('lowest plans')).toBe(true);
    expect(isPricingInstruction('vendor subscription')).toBe(true);
    expect(isPricingInstruction('per month')).toBe(true);
    expect(isPricingInstruction('per-month')).toBe(true);
    expect(isPricingInstruction('under $20')).toBe(true);
    expect(isPricingInstruction('approve if less than $5')).toBe(true);
  });

  it('does NOT match non-pricing questions', () => {
    expect(isPricingInstruction('is this a real token mint')).toBe(false);
    expect(isPricingInstruction('what protocol does jupiter use')).toBe(false);
    expect(isPricingInstruction('show recent transactions')).toBe(false);
    expect(isPricingInstruction('')).toBe(false);
    expect(isPricingInstruction('   ')).toBe(false);
  });
});

describe('filterLowAuthorityCitations', () => {
  const pricingInstruction = 'check helium mobile. lowest monthly plan. if less than $20. approve.';
  const nonPricingInstruction = 'is this a real token mint';

  function cite(url: string, title?: string) {
    return title ? { url, title } : { url };
  }

  it('drops blog/news/medium/substack citations on pricing questions', () => {
    const filtered = filterLowAuthorityCitations(
      [
        cite('https://blog.heliummobile.com/break-free', 'Break Free'),
        cite('https://news.vendor.com/article'),
        cite('https://medium.com/@user/post'),
        cite('https://author.substack.com/p/article'),
        cite('https://community.vendor.com/thread'),
        cite('https://forum.vendor.com/thread'),
        cite('https://user.wordpress.com/post'),
      ],
      pricingInstruction,
    );
    expect(filtered).toEqual([]);
  });

  it('preserves official pricing pages and support subdomains', () => {
    const filtered = filterLowAuthorityCitations(
      [
        cite('https://www.heliummobile.com/plans', 'Plans'),
        cite('https://hellohelium.com/plans'),
        cite('https://support.hellohelium.com/faq'),
        cite('https://pricing.vendor.com/tiers'),
        cite('https://vendor.com/about'),
      ],
      pricingInstruction,
    );
    expect(filtered).toHaveLength(5);
    expect(filtered[0]!.url).toBe('https://www.heliummobile.com/plans');
  });

  it('is a no-op for non-pricing questions (blog/medium pass through)', () => {
    const input = [
      cite('https://blog.example.com/post'),
      cite('https://medium.com/@user/post'),
      cite('https://example.com/api'),
    ];
    const filtered = filterLowAuthorityCitations(input, nonPricingInstruction);
    expect(filtered).toHaveLength(3);
  });

  it('drops a mixed list down to only official-domain citations', () => {
    const filtered = filterLowAuthorityCitations(
      [
        cite('https://blog.heliummobile.com/zero-plan'),
        cite('https://www.heliummobile.com/plans', 'Plans'),
        cite('https://hellohelium.com/'),
      ],
      pricingInstruction,
    );
    expect(filtered.map((c) => c.url)).toEqual([
      'https://www.heliummobile.com/plans',
      'https://hellohelium.com/',
    ]);
  });

  it('tolerates malformed URLs without crashing (does NOT drop them)', () => {
    const filtered = filterLowAuthorityCitations(
      [cite('not-a-url'), cite('https://www.heliummobile.com/plans')],
      pricingInstruction,
    );
    // Malformed entries pass through; we'd rather surface them than silently drop
    // potentially-legitimate citations on a parsing edge case.
    expect(filtered).toHaveLength(2);
  });
});

File: apps/android-twa/app/src/test/java/com/agentic/wallet/agent/provider/CitationFilterTest.kt
package com.agentic.wallet.agent.provider

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class CitationFilterTest {
    private val pricingInstruction =
        "check helium mobile. lowest monthly plan. if less than \$20. approve."
    private val nonPricingInstruction = "is this a real token mint"

    private fun cite(url: String, title: String? = null) = Citation(url, title)

    @Test
    fun pricingKeywordsAreDetected() {
        assertTrue(CitationFilter.isPricingInstruction("what is the price of helium mobile"))
        assertTrue(CitationFilter.isPricingInstruction("check the monthly cost"))
        assertTrue(CitationFilter.isPricingInstruction("lowest fee"))
        assertTrue(CitationFilter.isPricingInstruction("current rate"))
        assertTrue(CitationFilter.isPricingInstruction("cheapest plan"))
        assertTrue(CitationFilter.isPricingInstruction("lowest plans"))
        assertTrue(CitationFilter.isPricingInstruction("vendor subscription"))
        assertTrue(CitationFilter.isPricingInstruction("per month"))
        assertTrue(CitationFilter.isPricingInstruction("per-month"))
        assertTrue(CitationFilter.isPricingInstruction("under \$20"))
        assertTrue(CitationFilter.isPricingInstruction("approve if less than \$5"))
    }

    @Test
    fun nonPricingTextDoesNotMatch() {
        assertFalse(CitationFilter.isPricingInstruction("is this a real token mint"))
        assertFalse(CitationFilter.isPricingInstruction("what protocol does jupiter use"))
        assertFalse(CitationFilter.isPricingInstruction("show recent transactions"))
        assertFalse(CitationFilter.isPricingInstruction(""))
        assertFalse(CitationFilter.isPricingInstruction("   "))
        assertFalse(CitationFilter.isPricingInstruction(null))
    }

    @Test
    fun dropsBlogNewsMediumSubstackOnPricingQuestions() {
        val filtered = CitationFilter.filterLowAuthorityCitations(
            listOf(
                cite("https://blog.heliummobile.com/break-free", "Break Free"),
                cite("https://news.vendor.com/article"),
                cite("https://medium.com/@user/post"),
                cite("https://author.substack.com/p/article"),
                cite("https://community.vendor.com/thread"),
                cite("https://forum.vendor.com/thread"),
                cite("https://user.wordpress.com/post"),
            ),
            pricingInstruction,
        )
        assertEquals(emptyList<Citation>(), filtered)
    }

    @Test
    fun preservesOfficialPricingPagesAndSupportSubdomains() {
        val filtered = CitationFilter.filterLowAuthorityCitations(
            listOf(
                cite("https://www.heliummobile.com/plans", "Plans"),
                cite("https://hellohelium.com/plans"),
                cite("https://support.hellohelium.com/faq"),
                cite("https://pricing.vendor.com/tiers"),
                cite("https://vendor.com/about"),
            ),
            pricingInstruction,
        )
        assertEquals(5, filtered.size)
        assertEquals("https://www.heliummobile.com/plans", filtered[0].url)
    }

    @Test
    fun nonPricingQuestionsArePassThrough() {
        val input = listOf(
            cite("https://blog.example.com/post"),
            cite("https://medium.com/@user/post"),
            cite("https://example.com/api"),
        )
        val filtered = CitationFilter.filterLowAuthorityCitations(input, nonPricingInstruction)
        assertEquals(3, filtered.size)
    }

    @Test
    fun mixedListIsTrimmedToOfficialDomains() {
        val filtered = CitationFilter.filterLowAuthorityCitations(
            listOf(
                cite("https://blog.heliummobile.com/zero-plan"),
                cite("https://www.heliummobile.com/plans", "Plans"),
                cite("https://hellohelium.com/"),
            ),
            pricingInstruction,
        )
        assertEquals(
            listOf("https://www.heliummobile.com/plans", "https://hellohelium.com/"),
            filtered.map { it.url },
        )
    }

    @Test
    fun malformedUrlsArePreserved() {
        val filtered = CitationFilter.filterLowAuthorityCitations(
            listOf(cite("not-a-url"), cite("https://www.heliummobile.com/plans")),
            pricingInstruction,
        )
        // Malformed entries pass through; we'd rather surface them than silently drop
        // potentially-legitimate citations on a parsing edge case.
        assertEquals(2, filtered.size)
    }
}
